src/hooks/useDateFilter.tsx

```diff
@@ -1,94 +1,117 @@
- * and display an error message.
- */
-
-export const useDateFilter = (dateFilterProps: DateFilterHookPropsType) => {
-  if (dateFilterProps.inputRefs.length !== 2) {
-    console.warn(
-      "useDateFilter hook requires inputRefs to contain two refs. Unexpected behavior may ensue."
-    )
-  }
-  const { dateBefore, dateAfter } = dateFilterProps
-  const [displayDateRangeError, setDisplayDateRangeError] = useState("")
-
-  const invalidYearFormat = rangeContainsInvalidYearFormat(
-    dateAfter,
-    dateBefore
-  )
-  const invalidRange = endDateInvalid(dateAfter, dateBefore)
-
-  const invalid = invalidRange || invalidYearFormat
-  const invalidReason = invalidRange
-    ? "Error: Start date must be earlier than end date."
-    : "Error: Years must be 4 digits"
-  const dateRangeError = invalid ? invalidReason : ""
-
-  const dateFilterWithHookProps = {
-    ...dateFilterProps,
-    displayDateRangeError,
+export interface DateErrorState {
+  from?: string
+  to?: string
+  range?: string
+}
+
+export const useDateFilter = (props: DateFilterHookPropsType) => {
+  const { inputRefs, dateFrom, dateTo } = props
+  const [dateError, setDateError] = useState<DateErrorState>({})
+  const debounceRef = useRef<number | null>(null)
+
+  const validateField = (
+    value: string,
+    fieldName: "from" | "to"
+  ): string | undefined => {
+    if (!value) return undefined
+    const digits = value.replace(/[^\d]/g, "")
+    if (digits.length < 4 || formatInvalid(value))
+      return `Please enter a valid '${fieldName}' date.`
+    return undefined
   }
 
-  const validateDateRange = () => {
-    // dateRangeError is either an empty string or an error message. If error message is present, we want to
-    // focus on the first dateForm input...
-    if (dateRangeError) {
-      dateFilterProps.inputRefs[0].current.focus()
+  // Debounced validation
+  useEffect(() => {
+    if (debounceRef.current) clearTimeout(debounceRef.current)
+
+    debounceRef.current = window.setTimeout(() => {
+      const errors: DateErrorState = {}
+      errors.from = validateField(dateFrom, "from")
+      errors.to = validateField(dateTo, "to")
+
+      if (
+        !errors.from &&
+        !errors.to &&
+        dateFrom &&
+        dateTo &&
+        rangeInvalid(dateFrom, dateTo)
+      ) {
+        errors.range = "End date must be later than start date."
+      }
+
+      setDateError(errors)
+      debounceRef.current = null
+    }, 600)
+
+    return () => {
+      if (debounceRef.current) clearTimeout(debounceRef.current)
+    }
+  }, [dateFrom, dateTo])
+
+  // Manual validation (Apply button)
+  const validateDates = (from = dateFrom, to = dateTo) => {
+    const errors: DateErrorState = {}
+    errors.from = validateField(from, "from")
+    errors.to = validateField(to, "to")
+
+    if (!errors.from && !errors.to && from && to && rangeInvalid(from, to)) {
+      errors.range = "End date must be later than start date."
```

Comment:
I noticed that a similar error message is also attached on line 139 in the advanced search page. Why does it need to be explicitly set here when the other error types don't seem to? If it does need to be referenced and set multiple times, I would save the string as a constant for consistency.